zcutil/fetch-params.py

```diff
@@ -0,0 +1,310 @@
+else:
+    logging.error(" %s is not currently supported.", HOST_OS)
+
+def is_ipfs_daemon_running():
+    #Due to corner cases of graceful ipfs daemon shutdown via CLI,
+    #we check the ipfs swarm to see if it has a valid network state.
+    #Future IPFS CLI updates should address this with `ipfs status`
+
+    if HOST_OS == 'Windows':
+        p1 = Popen(['ipfs.exe', "swarm", "addrs"], stderr=PIPE, stdout=PIPE)
+    else:
+        p1 = Popen(['ipfs', "swarm","addrs"], stderr=PIPE, stdout=PIPE)
+    stdout, stderr =  p1.communicate()
+
+    if p1.returncode == 1:
+        logging.error("IPFS Daemon is not running. Please start it and try again.")
+        return False
+    else:
+        return True
+
+def use_ipfs(filename):
+    logging.debug("Writing to: %s" % PARAMS_DIR + filename)
+    logging.info("Requesting %s from: /ipfs/%s" % (filename,PARAMS[filename]['ipfs_bhash']))
+    if HOST_OS == 'Windows':
+        p1 = Popen(['ipfs.exe', "get", "--output", PARAMS_DIR + filename, "/ipfs/" + PARAMS[filename]['ipfs_bhash']], stdout=PIPE)
+    else:
+        p1 = Popen(['ipfs', "get", "--output", PARAMS_DIR + filename, "/ipfs/" + PARAMS[filename]['ipfs_bhash']] , stdout=PIPE)
+    buffer = p1.communicate()[0]
+
+    if p1.returncode == 1:
+        logging.error(buffer)
+    
+    verify_ipfs(filename, DOWNLOADING)
+
+def use_https(filename):
+    path = PARAMS_DIR + filename + ".dl"
+    url = PARAMS_URL + filename
+    req = Request(url)
+    try:
+        resp = urlopen(req)
+    except HTTPError as e:
+        logging.exception('The server couldn\'t handle request.')
+        logging.exception('Error code: ', e.code)
+    except URLError as e:
+        logging.exception('Failed to reach server.')
+        logging.exception('Reason: ', e.reason)
+    else:
+        total_size = int(resp.headers['Content-Length'])
+        respData = resp.read()
+
+    with open(path, "wb") as handle:
+        lock(handle, LOCK_EX )
+        handle.write(respData)
+        unlock(handle)
+
+    size = os.path.getsize(path)
+    if size == total_size :
+        verify_file(path, PARAMS[filename]["sha256"], DOWNLOADING)
+    else:
+        logging.error("Downloaded file size did not match:", size, total_size)
+
+def download_file(filename, protocol):
+    if protocol == "HTTPS":
+        use_https(filename)
+    elif protocol == "IPFS":
+        use_ipfs(filename)
+    else:
+        logging.error("%s is not currently supported for retrieving Zcash params.\n", protocol)
+
+def verify_ipfs(filename, download_state):
+    #Generate the IPFS to verify, but don't actually add the file
+    if HOST_OS == 'Windows':
+        p2 = Popen(['ipfs.exe', "add", "--n", "--Q", PARAMS_DIR + filename], stdout=PIPE)
+    else:
+        p2 = Popen(['ipfs', "add", "--n", "--Q", PARAMS_DIR + filename], stdout=PIPE)
+    buffer = p2.communicate()[0]
+
+    if buffer.decode("utf-8").replace('\n','') != PARAMS[filename]['ipfs_bhash']:
```

Comment:
```suggestion
    if PARAMS[filename]['ipfs_bhash'] not in buffer.decode("utf-8").replace('\n',''):
```

`buffer.decode("utf-8").replace('\n','')` returns the string "added QmQ8E53Fpp1q1zXvsqscskaQXrgyqfac9b3AqLxFxCubAz sapling-output.params", 

while `PARAMS[filename]['ipfs_bhash']` just returns "QmQ8E53Fpp1q1zXvsqscskaQXrgyqfac9b3AqLxFxCubAz".

even though the hashes were equal and the download was successful, i was getting the misleading error message "Download failed: Multihash on sapling-output.params does NOT match."